### PR TITLE
don't use hardcoded paths

### LINKED ISSUE #21: Differing binary locations on hosts
I've been trying to get syncoid working between various hosts (and OS types), but the script seems to assume a standard binary layout across hosts. Unfortunately, some OSs have different conventions and package managers don't always make the same assumptions.

What about detecting binaries based on `$PATH`, using `which <executable>`?

A side benefit would reduce the need to modify the script itself on "non-standard" systems to keep git updates simpler.

### LINKED ISSUE #85: FreeBSD symlinks
The INSTALL file reads:
```
  Simplest path workaround is symlinks, eg:
  root@bsd:~# ln -s /usr/bin/lzop /usr/local/bin/lzop
```
It should be:
```
   ln -s /usr/local/bin/lzop  /usr/bin/lzop
```
Also needed
```
  ln -s /usr/local/bin/pv /usr/bin/pv
```
Moving all the FreeBSD notes to the FREEBSD.readme is cleaner -imho.

### PR DESCRIPTION
**Pull request by phreaker0**
This PR removes the hard coded paths for all binaries and executes them bare, relying on a correct set $PATH. For compatibility reasons and CRON all previous used binary locations are added PATH environment variable. So this should only break the execution on the remote hosts if the PATH isn't set correctly there?

@jimsalterjrs Is this enough for you or should I also figure out how to adapt the PATH on the remote as well (i wasn't very successful with first attempts)

Fixes #21 
Fixes #85

### CODE REVIEW

**Comment by redmop** [comment]
So what if we make the executables a variable. Like $zpool="/sbin/zpool" or some such thing (don't actually use the variable name, I just could not think of anything better atm).

**Comment by phreaker0** [comment]
@jimsalterjrs I just checked and i didn't rolled it out yet on my production systems, only on my machine. Will deploy it soon to ubuntu/debian machines and will report back after a while.

**Comment by phreaker0** [comment]
@redmop They currently are in a variable, but if the target host uses different locations for the binaries it won't work. This case should be fixed by this PR, but can also potentially break cases where the remote hasn't setuped a correct $PATH environment variable.

**Comment by redmop** [comment]
The idea behind that is it would be easier to alter the binary path if we put it in a variable. Then if we tried to find it, we just change the variable and we're done.

File: syncoid [hunk 1]
```diff
@@ -48,17 +48,20 @@ my $debug = $args{'debug'};
...
+
+my $pvcmd = 'pv';
+my $mbuffercmd = 'mbuffer';
+my $sudocmd = 'sudo';
 my $mbufferoptions = '-q -s 128k -m 16M 2>/dev/null';
-# currently using ls to check for file existence because we aren't depending on perl
+# currently using which to check for command existence because we aren't depending on perl
 # being present on remote machines.
-my $lscmd = '/bin/ls';
+my $whichcmd = 'which';
```
[the remark this answers is not in the document]
Comment (phreaker0) [reply syncoid]: will do, thanks!

**Comment by tomhoover** [comment]
You should also remove the hardcoded paths from lines 676-720 for the various compression methods.

**Comment by phreaker0** [comment]
@tomhoover thx

**Comment by tomhoover** [comment]
I've tested on both MacOS (Mojave) and Debian 9--works well on both.

**Comment by jimsalterjrs** [comment]
@phreaker0 did you say you've been using this in prod now without an explicitly set PATH in crontab? If so, on what distros?

**Comment by phreaker0** [comment]
@phreaker0 I only run sanoid via systemd timers. But I switched my workstation to use cron and removed the default PATH set by the distribution (ubuntu 18.04), will report back next week.

**Comment by phreaker0** [comment]
I apparently never reported back, it worked for me on ubuntu 18.04 with cron. I just also tested syncoid from 18.04 to FreeBSD 11.2 with that patch and it also worked.

**Comment by jimsalterjrs** [comment]
@phreaker0 will you rebase, please?

**Comment by phreaker0** [comment]
@jimsalterjrs done

**Comment by jimsalterjrs** [comment]
TY!

Now if only there was a way to fix the issue with the shebang still needing to point to the correct location of Perl...

**Comment by phreaker0** [comment]
@jimsalterjrs what about `#!/usr/bin/env perl`?

**Comment by jimsalterjrs** [comment]
Last time I tested, that did not work properly in a default cron environment on Ubuntu or FreeBSD.

**Comment by phreaker0** [comment]
true, if $PATH is not populated this won't work. So I guess there is no easy way then.

**Comment by redmop** [comment]
We can suggest that they set the PATH in crontab.

**Comment by phreaker0** [comment]
@redmop this would be the proper solution, but would break some existing setups

**Pull request merged**